### PR TITLE
[ISSUE #4029] add shenyu-client-api-doc part1 client module

### PR DESCRIPTION
**Pull request by 847850277**
Make sure that:

- [ ] You have read the [contribution guidelines](https://shenyu.apache.org/community/contributor-guide).
- [ ] You submit test cases (unit or integration tests) that back your changes.
- [ ] Your local test passed `./mvnw clean install -Dmaven.javadoc.skip=true`.

### CODE REVIEW

File: shenyu-client/shenyu-client-core/src/main/java/org/apache/shenyu/client/core/client/AbstractContextRefreshedEventListener.java [hunk 1]
```diff
@@ -108,7 +112,24 @@ public void onApplicationEvent(@NonNull final ContextRefreshedEvent event) {
         }
         publisher.publishEvent(buildURIRegisterDTO(context, beans));
         beans.forEach(this::handle);
+        Map<String, Object> apiModules = context.getBeansWithAnnotation(ApiModule.class);
+        apiModules.forEach(this::handleApiDoc);
     }
+
+    private void handleApiDoc(final String name, final Object classes) {
```
Comment (dragon-zhang) [inline shenyu-client/shenyu-client-core/src/main/java/org/apache/shenyu/client/core/client/AbstractContextRefreshedEventListener.java:119]: `Object classes` rename to `Object bean`.

File: shenyu-client/shenyu-client-core/src/main/java/org/apache/shenyu/client/core/client/AbstractContextRefreshedEventListener.java [hunk 2]
```diff
@@ -108,7 +112,24 @@ public void onApplicationEvent(@NonNull final ContextRefreshedEvent event) {
...
+            if (method.isAnnotationPresent(ApiDoc.class)) {
+                List<ApiDocRegisterDTO> apis = buildApiDocDTO(apiModuleClass, method);
+                for (ApiDocRegisterDTO apiDocRegisterDTO : apis) {
+                    publisher.publishEvent(apiDocRegisterDTO);
+                }
+            }
+        }
+    }
+
+    protected abstract List<ApiDocRegisterDTO> buildApiDocDTO(Class<?> clazz, Method method);
```
Comment (dragon-zhang) [inline shenyu-client/shenyu-client-core/src/main/java/org/apache/shenyu/client/core/client/AbstractContextRefreshedEventListener.java:132]: better to add `Object bean` param.

File: shenyu-client/shenyu-client-dubbo/shenyu-client-apache-dubbo/src/main/java/org/apache/shenyu/client/apache/dubbo/ApacheDubboServiceBeanListener.java [hunk 3]
```diff
@@ -62,7 +71,45 @@ public ApacheDubboServiceBeanListener(final PropertiesConfig clientConfig,
...
+        String value = shenyuDubboClient.value();
+        String apiPath = contextPath + superPath + value;
+        ApiDocRegisterDTO build = ApiDocRegisterDTO.builder()
+                .consume(ShenyuClientConstants.MEDIA_TYPE_ALL_VALUE)
+                .produce(ShenyuClientConstants.MEDIA_TYPE_ALL_VALUE)
+                .httpMethod(ApiHttpMethodEnum.NOT_HTTP.getValue())
+                .contextPath(contextPath)
+                .ext("{}")
+                .document("{}")
+                .version("v0.01")
```
Comment (dragon-zhang) [inline shenyu-client/shenyu-client-dubbo/shenyu-client-apache-dubbo/src/main/java/org/apache/shenyu/client/apache/dubbo/ApacheDubboServiceBeanListener.java:100]: In `dubbo`, maybe we can build version from `ServiceBean`.

File: shenyu-client/shenyu-client-dubbo/shenyu-client-alibaba-dubbo/src/main/java/org/apache/shenyu/client/alibaba/dubbo/AlibabaDubboServiceBeanListener.java [hunk 4]
```diff
@@ -60,6 +69,44 @@ public AlibabaDubboServiceBeanListener(final PropertiesConfig clientConfig,
...
+        String value = shenyuDubboClient.value();
+        String apiPath = contextPath + superPath + value;
+        ApiDocRegisterDTO build = ApiDocRegisterDTO.builder()
+                .consume(ShenyuClientConstants.MEDIA_TYPE_ALL_VALUE)
+                .produce(ShenyuClientConstants.MEDIA_TYPE_ALL_VALUE)
+                .httpMethod(ApiHttpMethodEnum.NOT_HTTP.getValue())
+                .contextPath(contextPath)
+                .ext("{}")
+                .document("{}")
+                .version("v0.01")
```
Comment (dragon-zhang) [inline shenyu-client/shenyu-client-dubbo/shenyu-client-alibaba-dubbo/src/main/java/org/apache/shenyu/client/alibaba/dubbo/AlibabaDubboServiceBeanListener.java:97]: In `dubbo`, maybe we can build version from `ServiceBean`.

File: shenyu-client/shenyu-client-dubbo/shenyu-client-alibaba-dubbo/src/main/java/org/apache/shenyu/client/alibaba/dubbo/AlibabaDubboServiceBeanListener.java [hunk 5]
```diff
@@ -60,6 +69,44 @@ public AlibabaDubboServiceBeanListener(final PropertiesConfig clientConfig,
...
+                .rpcType(RpcTypeEnum.DUBBO.getName())
+                .apiDesc(apiDesc)
+                .apiPath(apiPath)
+                .apiSource(1)
+                .state(1)
```
Comment (dragon-zhang) [inline shenyu-client/shenyu-client-dubbo/shenyu-client-alibaba-dubbo/src/main/java/org/apache/shenyu/client/alibaba/dubbo/AlibabaDubboServiceBeanListener.java:102]: `apiSource` and `state`, please use an enum.

File: shenyu-client/shenyu-client-api-docs-annotations/src/main/java/org/apache/shenyu/client/apidocs/annotations/ApiModule.java [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+@Target({ElementType.TYPE, ElementType.METHOD})
+@Retention(RetentionPolicy.RUNTIME)
+@Documented
+public @interface ApiModule {
+
+    /**
+     * value.
+     * @return value
+     */
+    String value();
```
Comment (dragon-zhang) [inline shenyu-client/shenyu-client-api-docs-annotations/src/main/java/org/apache/shenyu/client/apidocs/annotations/ApiModule.java:39]: maybe add `required` and `example` ?

File: shenyu-client/shenyu-client-http/shenyu-client-springcloud/src/main/java/org/apache/shenyu/client/springcloud/init/SpringCloudClientEventListener.java [hunk 7]
```diff
@@ -92,6 +106,123 @@ public SpringCloudClientEventListener(final PropertiesConfig clientConfig,
...
+        return list;
+    }
+
+    private Map<ApiHttpMethodEnum, Triple<String[], String[], String[]>> httpMethodConsumeProduceValueMap(final Method method) {
+        Map<ApiHttpMethodEnum, Triple<String[], String[], String[]>> map = Maps.newHashMap();
+        GetMapping getMapping = AnnotatedElementUtils.findMergedAnnotation(method, GetMapping.class);
+        if (Objects.nonNull(getMapping)) {
+            map.put(ApiHttpMethodEnum.GET, Triple.of(getMapping.consumes(), getMapping.produces(), getMapping.value()));
+            return map;
+        }
```
Comment (dragon-zhang) [inline shenyu-client/shenyu-client-http/shenyu-client-springcloud/src/main/java/org/apache/shenyu/client/springcloud/init/SpringCloudClientEventListener.java:202]: How about this ?
```java
RequestMapping mapping = AnnotatedElementUtils.findMergedAnnotation(method, RequestMapping.class);
for(RequestMethod m : mapping.method()) {
    map.put(ApiHttpMethodEnum.valueOf(m.name()), Triple.of(mapping.consumes(), mapping.produces(), mapping.value()));
}
```

File: shenyu-client/shenyu-client-http/shenyu-client-springmvc/src/main/java/org/apache/shenyu/client/springmvc/init/SpringMvcClientEventListener.java [hunk 8]
```diff
@@ -80,7 +94,124 @@ public SpringMvcClientEventListener(final PropertiesConfig clientConfig,
...
+        });
+        return list;
+    }
+
+    private Map<ApiHttpMethodEnum, Triple<String[], String[], String[]>> httpMethodConsumeProduceValueMap(final Method method) {
+        Map<ApiHttpMethodEnum, Triple<String[], String[], String[]>> map = Maps.newHashMap();
+        GetMapping getMapping = AnnotatedElementUtils.findMergedAnnotation(method, GetMapping.class);
+        if (Objects.nonNull(getMapping)) {
+            map.put(ApiHttpMethodEnum.GET, Triple.of(getMapping.consumes(), getMapping.produces(), getMapping.value()));
+            return map;
```
Comment (dragon-zhang) [inline shenyu-client/shenyu-client-http/shenyu-client-springmvc/src/main/java/org/apache/shenyu/client/springmvc/init/SpringMvcClientEventListener.java:190]: maybe do the same changes in `springcloud client`

File: shenyu-client/shenyu-client-sofa/src/main/java/org/apache/shenyu/client/sofa/SofaServiceEventListener.java [hunk 9]
```diff
@@ -151,6 +159,44 @@ public void onApplicationEvent(final ContextRefreshedEvent contextRefreshedEvent
...
+        if (Objects.isNull(shenyuSofaClient)) {
+            return Lists.newArrayList();
+        }
+        List<ApiDocRegisterDTO> list = Lists.newArrayList();
+        String value = shenyuSofaClient.value();
+        String apiPath = contextPath + superPath + value;
+        ApiDocRegisterDTO build = ApiDocRegisterDTO.builder()
+                .consume(ShenyuClientConstants.MEDIA_TYPE_ALL_VALUE)
+                .produce(ShenyuClientConstants.MEDIA_TYPE_ALL_VALUE)
+                .httpMethod(ApiHttpMethodEnum.NOT_HTTP.getValue())
```
Comment (dragon-zhang) [inline shenyu-client/shenyu-client-sofa/src/main/java/org/apache/shenyu/client/sofa/SofaServiceEventListener.java:183]: I have seen a lot of similar code. Can you use template method pattern to abstract another parent class? The new abstract class extends from `AbstractContextRefreshedEventListener`, and `SofaServiceEventListener` extends from the new abstract class.

File: shenyu-client/shenyu-client-core/src/main/java/org/apache/shenyu/client/core/client/AbstractContextRefreshedEventListener.java (hunk 1 above)
Comment (847850277) [reply shenyu-client/shenyu-client-core/src/main/java/org/apache/shenyu/client/core/client/AbstractContextRefreshedEventListener.java]: fixed

File: shenyu-client/shenyu-client-core/src/main/java/org/apache/shenyu/client/core/client/AbstractContextRefreshedEventListener.java (hunk 2 above)
Comment (847850277) [reply shenyu-client/shenyu-client-core/src/main/java/org/apache/shenyu/client/core/client/AbstractContextRefreshedEventListener.java]: fixed

File: shenyu-client/shenyu-client-dubbo/shenyu-client-apache-dubbo/src/main/java/org/apache/shenyu/client/apache/dubbo/ApacheDubboServiceBeanListener.java (hunk 3 above)
Comment (847850277) [reply shenyu-client/shenyu-client-dubbo/shenyu-client-apache-dubbo/src/main/java/org/apache/shenyu/client/apache/dubbo/ApacheDubboServiceBeanListener.java:103]: i try this code,but it didn't work
```java

if (apiModuleClass.isAnnotationPresent(Service.class)) {
    Service dubboService = apiModuleClass.getAnnotation(Service.class);
    apiVersion = dubboService.version();
} else {
    DubboService dubboService = apiModuleClass.getAnnotation(DubboService.class);
    apiVersion = dubboService.version();
}

```

File: shenyu-client/shenyu-client-dubbo/shenyu-client-alibaba-dubbo/src/main/java/org/apache/shenyu/client/alibaba/dubbo/AlibabaDubboServiceBeanListener.java (hunk 4 above)
Comment (847850277) [reply shenyu-client/shenyu-client-dubbo/shenyu-client-alibaba-dubbo/src/main/java/org/apache/shenyu/client/alibaba/dubbo/AlibabaDubboServiceBeanListener.java:100]: i try this code,but it didn't work
```java

if (apiModuleClass.isAnnotationPresent(Service.class)) {
    Service dubboService = apiModuleClass.getAnnotation(Service.class);
    apiVersion = dubboService.version();
} else {
    DubboService dubboService = apiModuleClass.getAnnotation(DubboService.class);
    apiVersion = dubboService.version();
}

```

File: shenyu-client/shenyu-client-dubbo/shenyu-client-alibaba-dubbo/src/main/java/org/apache/shenyu/client/alibaba/dubbo/AlibabaDubboServiceBeanListener.java (hunk 5 above)
Comment (847850277) [reply shenyu-client/shenyu-client-dubbo/shenyu-client-alibaba-dubbo/src/main/java/org/apache/shenyu/client/alibaba/dubbo/AlibabaDubboServiceBeanListener.java]: fixed

File: shenyu-client/shenyu-client-api-docs-annotations/src/main/java/org/apache/shenyu/client/apidocs/annotations/ApiModule.java (hunk 6 above)
Comment (847850277) [reply shenyu-client/shenyu-client-api-docs-annotations/src/main/java/org/apache/shenyu/client/apidocs/annotations/ApiModule.java:39]: i think required and example It is a parameter of the request response body. It should not appear in the ApiModule

File: shenyu-client/shenyu-client-http/shenyu-client-springcloud/src/main/java/org/apache/shenyu/client/springcloud/init/SpringCloudClientEventListener.java (hunk 7 above)
Comment (847850277) [reply shenyu-client/shenyu-client-http/shenyu-client-springcloud/src/main/java/org/apache/shenyu/client/springcloud/init/SpringCloudClientEventListener.java:206]: in SpringMvcClientEventListener or SpringCloudClientEventListener. this code in line 122. 
`RequestMapping` or `GetMapping` They are different annotation


```java
RequestMapping requestMapping = AnnotatedElementUtils.findMergedAnnotation(method, RequestMapping.class);
        if (Objects.nonNull(requestMapping)) {
            List<ApiDocRegisterDTO> list = Lists.newArrayList();
            String produce = requestMapping.produces().length == 0 ? ShenyuClientConstants.MEDIA_TYPE_ALL_VALUE : String.join(",", requestMapping.produces());
            String consume = requestMapping.consumes().length == 0 ? ShenyuClientConstants.MEDIA_TYPE_ALL_VALUE : String.join(",", requestMapping.consumes());
            String[] values = requestMapping.value();
            for (String value : values) {
                String apiPath = contextPath + superPath + value;
                RequestMethod[] requestMethods = requestMapping.method();
                if (requestMethods.length == 0) {
                    requestMethods = RequestMethod.values();
                }
                for (RequestMethod requestMethod : requestMethods) {
                    ApiDocRegisterDTO build = ApiDocRegisterDTO.builder()
                            .consume(consume)
                            .produce(produce)
                            .httpMethod(ApiHttpMethodEnum.getValueByName(String.valueOf(requestMethod)))
                            .contextPath(contextPath)
                            .ext("{}")
                            .document("{}")
                            .version("v0.01")
                            .rpcType(RpcTypeEnum.HTTP.getName())
                            .apiDesc(apiDesc)
                            .apiPath(apiPath)
                            .apiSource(ApiSourceEnum.ANNOTATION_GENERATION.getValue())
                            .state(ApiStateEnum.PUBLISHED.getState())
                            .apiOwner("admin")
                            .eventType(EventType.REGISTER)
                            .build();
                    list.add(build);
                }
            }
            return list;
        }

```

File: shenyu-client/shenyu-client-sofa/src/main/java/org/apache/shenyu/client/sofa/SofaServiceEventListener.java (hunk 9 above)
Comment (847850277) [reply shenyu-client/shenyu-client-sofa/src/main/java/org/apache/shenyu/client/sofa/SofaServiceEventListener.java:186]: i read and learn `template method`. and this code looks like `template method`.
```java
private void handleApiDoc(final String name, final Object bean) {
        Class<?> apiModuleClass = AopUtils.isAopProxy(bean) ? AopUtils.getTargetClass(bean) : bean.getClass();
        final Method[] methods = ReflectionUtils.getUniqueDeclaredMethods(apiModuleClass);
        for (Method method : methods) {
            if (method.isAnnotationPresent(ApiDoc.class)) {
                List<ApiDocRegisterDTO> apis = buildApiDocDTO(bean, method);
                for (ApiDocRegisterDTO apiDocRegisterDTO : apis) {
                    publisher.publishEvent(apiDocRegisterDTO);
                }
            }
        }
    }
```
Is it necessary to add another abstract class?

File: shenyu-client/shenyu-client-dubbo/shenyu-client-apache-dubbo/src/main/java/org/apache/shenyu/client/apache/dubbo/ApacheDubboServiceBeanListener.java (hunk 3 above)
Comment (dragon-zhang) [reply shenyu-client/shenyu-client-dubbo/shenyu-client-apache-dubbo/src/main/java/org/apache/shenyu/client/apache/dubbo/ApacheDubboServiceBeanListener.java:103]: I mean `org.apache.dubbo.config.spring.ServiceBean#getVersion`

File: shenyu-client/shenyu-client-sofa/src/main/java/org/apache/shenyu/client/sofa/SofaServiceEventListener.java (hunk 9 above)
Comment (dragon-zhang) [reply shenyu-client/shenyu-client-sofa/src/main/java/org/apache/shenyu/client/sofa/SofaServiceEventListener.java:186]: I think yes.

File: shenyu-client/shenyu-client-http/shenyu-client-springcloud/src/main/java/org/apache/shenyu/client/springcloud/init/SpringCloudClientEventListener.java (hunk 7 above)
Comment (dragon-zhang) [reply shenyu-client/shenyu-client-http/shenyu-client-springcloud/src/main/java/org/apache/shenyu/client/springcloud/init/SpringCloudClientEventListener.java:206]: Yes, but `GetMapping` is a derived annotation of `RequestMapping`, and all the attributes come from `RequestMapping`.

File: shenyu-client/shenyu-client-dubbo/shenyu-client-apache-dubbo/src/main/java/org/apache/shenyu/client/apache/dubbo/ApacheDubboServiceBeanListener.java (hunk 3 above)
Comment (847850277) [reply shenyu-client/shenyu-client-dubbo/shenyu-client-apache-dubbo/src/main/java/org/apache/shenyu/client/apache/dubbo/ApacheDubboServiceBeanListener.java:103]: in `alibaba dubbo` or `apache dubbo`. it's seem look like empty.

image

image

File: shenyu-client/shenyu-client-sofa/src/main/java/org/apache/shenyu/client/sofa/SofaServiceEventListener.java (hunk 9 above)
Comment (847850277) [reply shenyu-client/shenyu-client-sofa/src/main/java/org/apache/shenyu/client/sofa/SofaServiceEventListener.java:186]: > I think yes.

i will try

File: shenyu-client/shenyu-client-dubbo/shenyu-client-apache-dubbo/src/main/java/org/apache/shenyu/client/apache/dubbo/ApacheDubboServiceBeanListener.java (hunk 3 above)
Comment (dragon-zhang) [reply shenyu-client/shenyu-client-dubbo/shenyu-client-apache-dubbo/src/main/java/org/apache/shenyu/client/apache/dubbo/ApacheDubboServiceBeanListener.java:103]: try config it in `<dubbo:service version="1.0.0">` in xml, or `@DubboService(version="1.0.0")` in code.

Comment (847850277) [reply shenyu-client/shenyu-client-dubbo/shenyu-client-apache-dubbo/src/main/java/org/apache/shenyu/client/apache/dubbo/ApacheDubboServiceBeanListener.java]: i will try

File: shenyu-client/shenyu-client-sofa/src/main/java/org/apache/shenyu/client/sofa/SofaServiceEventListener.java (hunk 9 above)
Comment (847850277) [reply shenyu-client/shenyu-client-sofa/src/main/java/org/apache/shenyu/client/sofa/SofaServiceEventListener.java]: > I think yes.

fix it,The template method is used to design the pattern, but I don't think it is necessary to add new subclasses. Do you think the current implementation method is OK?

**Review by dragon-zhang — APPROVED** [review]

**Pull request merged**